XamarinCommunityToolkit/ObjectModel/ObservableRangeCollection.shared.cs

```diff
@@ -0,0 +1,164 @@
+	/// </summary>
+	/// <typeparam name="T"></typeparam>
+	public class ObservableRangeCollection<T> : ObservableCollection<T>
+	{
+
+		/// <summary>
+		/// Initializes a new instance of the System.Collections.ObjectModel.ObservableCollection(Of T) class.
+		/// </summary>
+		public ObservableRangeCollection()
+			: base()
+		{
+		}
+
+		/// <summary>
+		/// Initializes a new instance of the System.Collections.ObjectModel.ObservableCollection(Of T) class that contains elements copied from the specified collection.
+		/// </summary>
+		/// <param name="collection">collection: The collection from which the elements are copied.</param>
+		/// <exception cref="System.ArgumentNullException">The collection parameter cannot be null.</exception>
+		public ObservableRangeCollection(IEnumerable<T> collection)
+			: base(collection)
+		{
+		}
+
+		/// <summary>
+		/// Adds the elements of the specified collection to the end of the ObservableCollection(Of T).
+		/// </summary>
+		public void AddRange(IEnumerable<T> collection, NotifyCollectionChangedAction notificationMode = NotifyCollectionChangedAction.Add)
+		{
+			if (notificationMode != NotifyCollectionChangedAction.Add && notificationMode != NotifyCollectionChangedAction.Reset)
+				throw new ArgumentException("Mode must be either Add or Reset for AddRange.", nameof(notificationMode));
+			if (collection == null)
+				throw new ArgumentNullException(nameof(collection));
+
+			CheckReentrancy();
+
+			var startIndex = Count;
+
+			var itemsAdded = AddArrangeCore(collection);
+
+			if (!itemsAdded)
+				return;
+
+			if (notificationMode == NotifyCollectionChangedAction.Reset)
+			{
+				RaiseChangeNotificationEvents(action: NotifyCollectionChangedAction.Reset);
+
+				return;
+			}
+
+			var changedItems = collection is List<T> ? (List<T>)collection : new List<T>(collection);
+
+			RaiseChangeNotificationEvents(
+				action: NotifyCollectionChangedAction.Add,
+				changedItems: changedItems,
+				startingIndex: startIndex);
+		}
+
+		/// <summary>
+		/// Removes the first occurence of each item in the specified collection from ObservableCollection(Of T). NOTE: with notificationMode = Remove, removed items starting index is not set because items are not guaranteed to be consecutive.
+		/// </summary>
+		public void RemoveRange(IEnumerable<T> collection, NotifyCollectionChangedAction notificationMode = NotifyCollectionChangedAction.Reset)
+		{
+			if (notificationMode != NotifyCollectionChangedAction.Remove && notificationMode != NotifyCollectionChangedAction.Reset)
+				throw new ArgumentException("Mode must be either Remove or Reset for RemoveRange.", nameof(notificationMode));
+			if (collection == null)
+				throw new ArgumentNullException(nameof(collection));
+
+			CheckReentrancy();
+
+			if (notificationMode == NotifyCollectionChangedAction.Reset)
+			{
+				var raiseEvents = false;
+				foreach (var item in collection)
+				{
+					Items.Remove(item);
```

Comment:
Won't we get an exception if try to call RemoveRange on a collection with passing the same collection as an argument?

```
collection.RemoveRange(collection);
```